python/simulation.py

```diff
@@ -0,0 +1,689 @@
+            elif isinstance(s, Rotate4):
+                s.swigobj = mp.rotate4(s.direction, gv)
+                sym += s.swigobj * complex(s.phase.real, s.phase.imag)
+            elif isinstance(s, Mirror):
+                s.swigobj = mp.mirror(s.direction, gv)
+                sym += s.swigobj * complex(s.phase.real, s.phase.imag)
+            else:
+                s.swigobj = mp.symmetry()
+
+        br = _create_boundary_region_from_pml_layers(self.pml_layers, gv)
+
+        self.structure = mp.structure(gv, dummy_eps, br, sym, self.num_chunks, self.courant,
+                                      self.eps_averaging, self.subpixel_tol, self.subpixel_maxeval)
+
+        mp.set_materials_from_geometry(self.structure, self.geometry)
+
+    def _init_fields(self):
+        is_cylindrical = self.dimensions == CYLINDRICAL
+
+        if self.structure is None:
+            self._init_structure(self.k_point)
+
+        self.fields = mp.fields(
+            self.structure,
+            self.m if is_cylindrical else 0,
+            self.k_point.z if self.special_kz and self.k_point else 0,
+            not self.accurate_fields_near_cylorigin
+        )
+
+        if self.verbose:
+            self.fields.verbose()
+
+        def use_real(self):
+            cond1 = is_cylindrical and self.m != 0
+            cond2 = any([s.phase.imag for s in self.symmetries])
+            cond3 = not self.k_point
+            cond4 = self.special_kz and self.k_point.x == 0 and self.k_point.y == 0
+            cond5 = not (cond3 or cond4 or self.k_point == Vector3())
+            return not (self.force_complex_fields or cond1 or cond2 or cond5)
+
+        if use_real(self):
+            self.fields.use_real_fields()
+        else:
+            print("Meep: using complex fields.")
+
+        if self.k_point:
+            v = Vector3(self.k_point.x, self.k_point.y) if self.special_kz else self.k_point
+            self.fields.use_bloch(v)
+
+        for s in self.sources:
+            self.add_source(s)
+
+        for hook in self.init_fields_hooks:
+            hook()
+
+    def meep_time(self):
+        if self.fields is None:
+            self._init_fields()
+        return self.fields.time()
+
+    def _round_time(self):
+        if self.fields is None:
+            self._init_fields()
+
+        return self.fields.round_time()
+
+    def _get_field_point(self, c, pt):
+        v3 = py_v3_to_vec(self.dimensions, pt)
+        return self.fields.get_field_from_comp(c, v3)
+
+    def _get_filename_prefix(self):
+        if self.include_files and self.filename_prefix == '':
+            _, filename = os.path.split(self.include_files[0])
+            return re.sub(r'\.py', '', filename)
+        else:
+            return self.filename_prefix
+
+    def _eval_step_func(self, func, todo):
+        num_args = get_num_args(func)
+        if num_args == 0:
+            if todo == 'step':
+                func()
+        else:
+            func(todo)
```

Comment:
Unlike Scheme (where everything is a global), probably the step functions need to be passed `self` (the current `Simulation` object) so that they have access to the fields etcetera.